Replace cURL with PSR-7 and PSR-18 implementation

DanielRuf opened the pull request:
This PR implements the suggested features from #6.

DanielRuf committed:
Replace cURL with PSR-7 and PSR-18 implementation

shochdoerfer commented on composer.json:
```diff
@@ -10,6 +10,10 @@
     "require": {
-        "ext-curl": "*"
+        "psr/http-client": "^1.0",
+        "nyholm/psr7": "^1.1"
```
If we define the purpose of this project to be a library, we don't need a concrete implementation of PSR-7, we would simply depend on the generic interfaces.

shochdoerfer commented on testiat.php:
```diff
@@ -51,27 +69,29 @@ private function createRequest($queryArray, $path) {
-		}
-		return json_decode($response);
+
+        $psr17Factory = new \Nyholm\Psr7\Factory\Psr17Factory();
```
The factory, specifically the RequestFactoryInterface should be injected via constructor and not created here manually. That way the users could use whatever they want as an PSR-17 implementation.

shochdoerfer commented on testiat.php:
```diff
@@ -51,27 +69,29 @@ private function createRequest($queryArray, $path) {
+        $request = ($psr17Factory->createRequest(
+            'POST',
+            self::API_ENPOINT . $path
+        ))->withBody($body);
```
You could omit passing the body instance and just use the ->getBody() call on $request to rewind the stream first and then write to it. It makes sense to rewind the stream first to be sure that the request body contains only the data you define.

shochdoerfer commented on testiat.php:
```diff
@@ -18,7 +33,10 @@ class Api
+    ) {
+        $this->client = $client;
 
         echo self::INTRO;
```
We should get rid of the echo calls as this violates SRP - the class should only work as an Http client implementation and not echo stuff on the console. If we really need cli interaction we could make use of the Symfony CLI package.

shochdoerfer commented on testiat.php:
```diff
@@ -85,7 +105,7 @@ public function getProjectStatus($id) {
     }
 
-    public function startEmailTest($subject, $html, $clients) {
+    public function startEmailTest($subject, $html, $clients): array {
```
Ideally we should also make use of method parameter type hints.

shochdoerfer commented on testiat.php:
```diff
@@ -51,27 +69,29 @@ private function createRequest($queryArray, $path) {
+            return false;
+        }
+
+        return json_decode((string) $response->getBody(), true);
```
Depending on what we want to achieve it might make sense to just return the response object instead of returning just the JSON output.

DanielRuf commented:
@shochdoerfer thanks for the review, I'll address the feedback this week.

DanielRuf committed:
Require only the generic http-client interfaces

DanielRuf committed:
Apply changes from PR review

Remove echo statements

Inject RequestFactoryInterface via constructor

Add parameter type hints

DanielRuf committed:
Remove obsolete code and return response object

DanielRuf committed:
Reformat example

DanielRuf commented:
Bump =)
Let me know if there is anything that I oversaw or did wrong.

DanielRuf committed:
Use single quotes

shochdoerfer committed:
Introduce proper namespace & autoloading support

shochdoerfer committed:
Add missing Composer dependencies

shochdoerfer committed:
Make code PSR-2 compatible

shochdoerfer committed:
Restructure class

shochdoerfer committed:
Replace self:: calls with $this

shochdoerfer committed:
Refactor createRequest logic

shochdoerfer committed:
Add docblock comments

shochdoerfer commented:
@DanielRuf applied some refactorings. Ready not to be merged.

DanielRuf commented on src/bitExpert/Testiat/Api.php:
```diff
@@ -0,0 +1,105 @@
+        try {
+            $request = $this->factory->createRequest(
+                'POST',
+                self::API_ENPOINT . $path
```
Do we have to use $this here too?

shochdoerfer replied:
No, because the constant is kind of globally accessible.

shochdoerfer commented:
Depends, we could log it, but then we need an additional dependency. For me it's fine to ignore it.

DanielRuf commented:
> @DanielRuf applied some refactorings. Ready not to be merged.

Awesome =) :+1: Do you mean "now"?

shochdoerfer commented:
Sure ;)

DanielRuf committed:
Merge branch 'master' into refactor/psr7-psr18-support
